src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,812 @@
+   * @param _tokenIds          Array of Token IDs that are sold
+   * @param _tokensSoldAmounts Array of Amount of Tokens sold for each id in _tokenIds.
+   * @param _minCurrency       Minimum amount of currency tokens to receive
+   * @param _deadline          Timestamp after which this transaction will be reverted
+   * @param _recipient         The address that receives output currency tokens.
+   * @return currencyBought How much currency was actually purchased.
+   */
+  function _tokenToCurrency(
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokensSoldAmounts,
+    uint256 _minCurrency,
+    uint256 _deadline,
+    address _recipient)
+    internal nonReentrant() returns (uint256[] memory currencyBought)
+  {
+    // Number of Token IDs to deposit
+    uint256 nTokens = _tokenIds.length;
+
+    // Input validation
+    require(_deadline >= block.timestamp, "NiftyswapExchange#_tokenToCurrency: DEADLINE_EXCEEDED");
+
+    // Initialize variables
+    uint256 totalCurrency = 0; // Total amount of currency tokens to transfer
+    currencyBought = new uint256[](nTokens);
+    uint256[] memory tokenReserves = new uint256[](nTokens);
+
+    // Get token reserves
+    tokenReserves = _getTokenReserves(_tokenIds);
+
+    // Assumes the Tokens ids are already received by contract, but not
+    // the Tokens Ids. Will return cards not sold if invalid price.
+
+    // Remove liquidity for each Token ID in _tokenIds
+    for (uint256 i = 0; i < nTokens; i++) {
+      // Store current id and amount from argument arrays
+      uint256 idSold = _tokenIds[i];
+      uint256 amountSold = _tokensSoldAmounts[i];
+      uint256 tokenReserve = tokenReserves[i];
+
+      // If 0 tokens send for this ID, revert
+      require(amountSold > 0, "NiftyswapExchange#_tokenToCurrency: NULL_TOKENS_SOLD");
+
+      // Load currency token and Token _id reserves
+      uint256 currencyReserve = currencyReserves[idSold];
+
+      // Get amount of currency that will be received
+      // Need to sub amountSold because tokens already added in reserve, which would bias the calculation
+      // Don't need to add it for currencyReserve because the amount is added after this calculation
+      uint256 currencyAmount = getSellPrice(amountSold, tokenReserve.sub(amountSold), currencyReserve);
+
+      // Increase cost of transaction
+      totalCurrency = totalCurrency.add(currencyAmount);
+
+      // Update individual currency reseve amount
+      currencyReserves[idSold] = currencyReserve.sub(currencyAmount);
+
+      // Append Token id, Token id amount and currency token amount to tracking arrays
+      currencyBought[i] = currencyAmount;
+    }
+
+    // If minCurrency is not met
+    require(totalCurrency >= _minCurrency, "NiftyswapExchange#_tokenToCurrency: INSUFFICIENT_CURRENCY_AMOUNT");
+
+    // Transfer currency here
+    currency.transfer(_recipient, totalCurrency);
```

Comment:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.